### PR TITLE
Add new stateless FK

### PR DESCRIPTION
**Pull request by exhaustin**
## Changes
- Add recursive implementation of forward kinematics
- Add forward kinematics API that returns poses off all links (instead of just the specified end-effector), which uses the recursive implementation

### CODE REVIEW

File: differentiable_robot_model/robot_model.py [hunk 1]
```diff
@@ -128,6 +128,30 @@ def __init__(self, urdf_path: str, name="", device=None):
             self._bodies.append(body)
             self._name_to_idx_map[body.name] = i
 
+        # New way of maintaining relationships between bodies
+        self._bodies2 = {}
```
Comment (fmeier) [inline differentiable_robot_model/robot_model.py:132]: how is this _bodies2 different from _bodies ? Why do we need both?

File: differentiable_robot_model/robot_model.py [hunk 2]
```diff
@@ -633,46 +688,52 @@ def _get_parent_object_of_param(self, link_name: str, parameter_name: str):
         body_idx = self._name_to_idx_map[link_name]
         if parameter_name in ["trans", "rot_angles", "joint_damping"]:
             parent_object = self._bodies[body_idx]
+            parent_object2 = self._bodies2[link_name]
```
Comment (fmeier) [inline differentiable_robot_model/robot_model.py:691]: @exhaustin I don't fully understand what's happening here. It looks like we return the parent twice - why?

File: differentiable_robot_model/rigid_body.py [hunk 3]
```diff
@@ -66,8 +74,54 @@ def __init__(self, rigid_body_params, device="cpu"):
...
-        return
+    # Kinematic tree construction
+    def set_parent(self, link: "DifferentiableRigidBody"):
+        self._parent = link
+
+    def add_child(self, link: "DifferentiableRigidBody"):
+        self._children.append(link)
+
+    # Recursive algorithms
+    def forward_kinematics(self, q_dict):
```
Comment (fmeier) [inline differentiable_robot_model/rigid_body.py:85]: can we add a comment about what this function does exactly?

**Review by fmeier — COMMENTED** [review]
I have added a few questions

**Comment by exhaustin** [comment]
Note: Previous confusing changes involving `self._bodies2` and various duplicated code was for indexing bodies with link names instead of body indices.
I cleaned up by getting it to work with the original indexing method in `self._bodies`.

**Review by fmeier — APPROVED** [review]
lgtm!

**Pull request merged**